File: android/src/oldarch/ReactNativeBrownfieldModule.kt
package com.callstack.reactnativebrownfield

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

/**
 * Callback interface for CTA button press events from React Native
 */
interface CTACallback {
  fun onCTAPressed(action: String)
}

/**
 * Callback interface for debug log messages from React Native
 */
interface DebugLogCallback {
  fun onDebugLog(level: String, message: String, context: String?, timestamp: Double)
}

/**
 * Callback interface for analytics events from React Native
 */
interface AnalyticsEventCallback {
  fun onAnalyticsEvent(eventName: String, eventProperties: Map<String, Any>)
}

class ReactNativeBrownfieldModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {
  companion object {
    var shouldPopToNative: Boolean = false
    private var moduleInstance: ReactNativeBrownfieldModule? = null
    
    /**
     * Callback listener for CTA events from React Native
     */
    @JvmStatic
    var ctaCallback: CTACallback? = null

    /**
     * Callback listener for debug log messages from React Native
     */
    @JvmStatic
    var debugLogCallback: DebugLogCallback? = null

    /**
     * Callback listener for analytics events from React Native
     */
    @JvmStatic
    var analyticsEventCallback: AnalyticsEventCallback? = null

    /**
     * Send GSM device status update to React Native
     * @param isConnected Boolean indicating if GSM device is connected
     */
    @JvmStatic
    fun notifyGSMStatusChanged(isConnected: Boolean) {
      moduleInstance?.sendGSMStatusEvent(isConnected)
    }
  }

  init {
    moduleInstance = this
  }

  @ReactMethod
  fun addListener(eventName: String) {
    // Required for RCTEventEmitter compatibility
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    // Required for RCTEventEmitter compatibility
  }

  private fun sendGSMStatusEvent(isConnected: Boolean) {
    reactApplicationContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      ?.emit("onGSMDeviceStatusChanged", createGSMStatusMap(isConnected))
  }

  private fun createGSMStatusMap(isConnected: Boolean): WritableMap {
    return Arguments.createMap().apply {
      putBoolean("connected", isConnected)
    }
  }

  @ReactMethod
  fun popToNative(animated: Boolean) {
    shouldPopToNative = true
    onBackPressed()
  }

  @ReactMethod
  fun setPopGestureRecognizerEnabled(enabled: Boolean) {
    shouldPopToNative = enabled
  }

  @ReactMethod
  fun setHardwareBackButtonEnabled(isFirstRoute: Boolean) {
    shouldPopToNative = isFirstRoute
  }

  @ReactMethod
  fun onCTAPressed(action: String) {
    reactApplicationContext.currentActivity?.runOnUiThread {
      ctaCallback?.onCTAPressed(action)
    }
  }

  @ReactMethod
  fun sendDebugLog(level: String, message: String, context: String?, timestamp: Double) {
    reactApplicationContext.currentActivity?.runOnUiThread {
      debugLogCallback?.onDebugLog(level, message, context, timestamp)
    }
  }

  @ReactMethod
  fun sendAnalyticsEvent(eventName: String, eventProperties: String) {
    reactApplicationContext.currentActivity?.runOnUiThread {
      try {
        // Parse JSON string to Map
        val gson = com.google.gson.Gson()
        val mapType = object : com.google.gson.reflect.TypeToken<Map<String, Any>>() {}.type
        val propertiesMap: Map<String, Any> = gson.fromJson(eventProperties, mapType)
        analyticsEventCallback?.onAnalyticsEvent(eventName, propertiesMap)
      } catch (e: Exception) {
        // Fallback to empty map if JSON parsing fails
        analyticsEventCallback?.onAnalyticsEvent(eventName, emptyMap())
      }
    }
  }

  private fun onBackPressed() {
    reactApplicationContext.currentActivity?.runOnUiThread {
      reactApplicationContext.currentActivity?.onBackPressed()
    }
  }

  override fun getName(): String {
    return "ReactNativeBrownfield"
  }
}

File: ios/ReactNativeBrownfieldModule.swift
internal import React

@objcMembers
public class ReactNativeBrownfieldModuleImpl: NSObject {
  static public func setPopGestureRecognizerEnabled(_ enabled: Bool) {
    let userInfo = ["enabled": enabled]
    DispatchQueue.main.async {
      NotificationCenter.default.post(name: Notification.Name.togglePopGestureRecognizer, object: nil, userInfo: userInfo)
    }
  }

  static public func popToNative(animated: Bool) {
    let userInfo = ["animated": animated]
    DispatchQueue.main.async {
      NotificationCenter.default.post(name: Notification.Name.popToNative, object: nil, userInfo: userInfo)
    }
  }
  
  static public func notifyGSMStatusChanged(_ isConnected: Bool) {
    let userInfo = ["connected": isConnected]
    DispatchQueue.main.async {
      NotificationCenter.default.post(
        name: NSNotification.Name.gsmDeviceStatusChanged,
        object: nil,
        userInfo: userInfo
      )
    }
  }

  static public func onCTAPressed(action: String) {
    let userInfo = ["action": action]
    DispatchQueue.main.async {
      NotificationCenter.default.post(
        name: NSNotification.Name.ctaPressed,
        object: nil,
        userInfo: userInfo
      )
    }
  }

  static public func sendDebugLog(level: String, message: String, context: String?, timestamp: Double) {
    var userInfo: [String: Any] = [
      "level": level,
      "message": message,
      "timestamp": timestamp
    ]
    if let context = context {
      userInfo["context"] = context
    }
    DispatchQueue.main.async {
      NotificationCenter.default.post(
        name: NSNotification.Name.debugLogReceived,
        object: nil,
        userInfo: userInfo
      )
    }
  }

  static public func sendAnalyticsEvent(eventName: String, eventProperties: String) {
    var userInfo: [String: Any] = [
      "eventName": eventName
    ]
    
    // Parse JSON string to Dictionary
    if let data = eventProperties.data(using: .utf8) {
      do {
        if let propertiesDict = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
          userInfo["eventProperties"] = propertiesDict
        } else {
          userInfo["eventProperties"] = [:]
        }
      } catch {
        userInfo["eventProperties"] = [:]
      }
    } else {
      userInfo["eventProperties"] = [:]
    }
    
    DispatchQueue.main.async {
      NotificationCenter.default.post(
        name: NSNotification.Name.analyticsEventReceived,
        object: nil,
        userInfo: userInfo
      )
    }
  }
}
